### PR TITLE
Correctly set imageset_id column in integration

### PR DESCRIPTION
**Pull request by jbeilstenedmands**
Currently, the `imageset_id` column is only output by `dials.integrate` if `include_bad_reference=True` (which we turn on by default for `xia2`). Fortunately the result is accidentally correct for integrating a single-sweep (due to the default null value of flex.int being 0) but wrong for multi sweep data.

The fix is to actually set the `imageset_id` at the prediction stage of `dials.integrate`. 

To reproduce the bug, take any multi-sweep data and import together in dials.import
```
dials.import sweep1*.cbf sweep2*.cbf
dials.find_spots imported.expt
dials.index imported.expt strong.refl
dials.integrate indexed.* include_bad_reference = True
```
Now if you open the integrated files in the `dials.image_viewer`, currently all integrated reflections will be shown on the first imageset, and any unintegrated reflections are shown correctly on their imagesets. If `include_bad_reference=False` (default for `dials.integrate`) then the image viewer correctly works out which reflections belong to which image through the `id` column.

### CODE REVIEW

File: command_line/integrate.py [hunk 1]
```diff
@@ -482,6 +484,17 @@ def run_integration(params, experiments, reference=None):
         force_static=params.prediction.force_static,
         padding=params.prediction.padding,
     )
+    isets = OrderedSet([e.imageset for e in experiments])
+    predicted["imageset_id"] = flex.int(predicted.size(), 0)
+    if len(list(isets)) > 1:
+        for e in experiments:
+            iset_id = list(isets).index(e.imageset)
```
Comment (Anthchirp) [inline command_line/integrate.py:491]:
```suggestion
            iset_id = isets.index(e.imageset)
```

File: command_line/integrate.py [hunk 2]
```diff
@@ -482,6 +484,17 @@ def run_integration(params, experiments, reference=None):
         force_static=params.prediction.force_static,
         padding=params.prediction.padding,
     )
+    isets = OrderedSet([e.imageset for e in experiments])
```
Comment (Anthchirp) [inline command_line/integrate.py:487]:
```suggestion
    isets = OrderedSet(e.imageset for e in experiments)
```

File: command_line/integrate.py [hunk 3]
```diff
@@ -482,6 +484,17 @@ def run_integration(params, experiments, reference=None):
...
+    predicted["imageset_id"] = flex.int(predicted.size(), 0)
+    if len(list(isets)) > 1:
```
Comment (Anthchirp) [inline command_line/integrate.py:489]:
```suggestion
    if len(isets) > 1:
```

File: tests/command_line/test_integrate.py [hunk 4]
```diff
@@ -254,6 +254,71 @@ def test_integration_with_sample_size(dials_data, tmpdir):
     assert dict(table.experiment_identifiers()) == {0: "foo"}
 
 
+def test_imageset_id_output_with_multi_sweep(dials_data, tmpdir):
+    """Test that imageset ids are correctly output for multi-sweep integration."""
+    # Just integrate 15 images for each sweep
+
+    images1 = dials_data("l_cysteine_dials_output") / "l-cyst_01_000*.cbf"
+    images2 = dials_data("l_cysteine_dials_output") / "l-cyst_02_000*.cbf"
```
Comment (Anthchirp) [inline tests/command_line/test_integrate.py:262]:
```suggestion
def test_imageset_id_output_with_multi_sweep(dials_data, tmp_path):
    """Test that imageset ids are correctly output for multi-sweep integration."""
    # Just integrate 15 images for each sweep

    images1 = dials_data("l_cysteine_dials_output", pathlib=True) / "l-cyst_01_000*.cbf"
    images2 = dials_data("l_cysteine_dials_output", pathlib=True) / "l-cyst_02_000*.cbf"
```

and a blanket search/replace `tmpdir` -> `tmp_path` in the function

File: tests/command_line/test_integrate.py [hunk 5]
```diff
@@ -289,7 +289,7 @@ def test_imageset_id_output_with_multi_sweep(dials_data, tmp_path):
     )
     assert not result.returncode and not result.stderr
     table = flex.reflection_table.from_file(tmp_path / "integrated.refl")
-    assert len(set(table["imageset_id"])) == 2
+    assert set(table["imageset_id"]) == set([0, 1])
```
Comment (Anthchirp) [inline tests/command_line/test_integrate.py:292]:
```suggestion
    assert set(table["imageset_id"]) == {0, 1}
```

File: tests/command_line/test_integrate.py [hunk 6]
```diff
@@ -312,7 +312,7 @@ def test_imageset_id_output_with_multi_sweep(dials_data, tmp_path):
...
```
Comment (Anthchirp) [inline tests/command_line/test_integrate.py:315]:
```suggestion
    assert set(table["imageset_id"]) == {0, 1}
```

File: tests/command_line/test_integrate.py [hunk 7]
```diff
@@ -443,7 +443,7 @@ def test_multi_lattice(dials_regression, tmpdir):
     assert len(table) == 5605
     assert dict(table.experiment_identifiers()) == {0: "100", 1: "101"}
     # both should only have the imageset_id of zero as they share an imageset
-    assert len(set(table["imageset_id"])) == 1
+    assert set(table["imageset_id"]) == set([0])
```
Comment (Anthchirp) [inline tests/command_line/test_integrate.py:446]:
```suggestion
    assert set(table["imageset_id"]) == {0}
```

**Review by Anthchirp — COMMENTED** [review]
cf. #1764

**Comment by ndevenish** [comment]
This appears to not update any unintegrated reflections - is this information just permanently lost, at the moment?

**Comment by ndevenish** [comment]
After discussion, we concluded that this is fine and I misinterpreted the wider context of the code.

**Pull request merged**